include/vulkan/vulkan_core.h

```diff
@@ -46,15 +46,36 @@ extern "C" {
 #define VK_HEADER_VERSION 82
 
 
-#define VK_NULL_HANDLE 0
+#ifdef __cplusplus
+    struct Null_t{ explicit operator uint64_t() const { return 0; } };
+    #define VK_NULL_HANDLE Null_t{}
+#else
+    #define VK_NULL_HANDLE 0
+#endif
         
 
 
 #define VK_DEFINE_HANDLE(object) typedef struct object##_T* object;
 
 
 #if !defined(VK_DEFINE_NON_DISPATCHABLE_HANDLE)
-#if defined(__LP64__) || defined(_WIN64) || (defined(__x86_64__) && !defined(__ILP32__) ) || defined(_M_X64) || defined(__ia64) || defined (_M_IA64) || defined(__aarch64__) || defined(__powerpc64__)
+#ifdef __cplusplus
+        #define VK_DEFINE_NON_DISPATCHABLE_HANDLE(object) \
+            class object{ \
+            private: \
+                uint64_t h; \
+            public: \
+                object(){}; \
+                object( const Null_t& nh ){ h = static_cast<uint64_t>(nh); } \
+                explicit object( const uint64_t rh ){ h = rh; } \
+                operator bool() const{ return this->h; } \
```

Comment:
Would it be better for this to be `explicit`? Implicit conversion to bool can sometimes be very surprising.

Reply:
Personally, I love to be able to do `if( handle ) dosomething();`. So implicit conversion to **some** _integral type_ has to be there to allow it, right?

Reply:
Actually, [`explicit bool` works in that case](https://godbolt.org/g/tAhq9K). See [discussion of the problem](https://en.cppreference.com/w/cpp/language/implicit_conversion#The_safe_bool_problem).

Reply:
thanks, seems so, it is C++11 though. Not sure how much should I worry about C++98\03. I can probably guard the whole thing for C++11, and people using obsolete compilers can just get the C version...

Reply:
done